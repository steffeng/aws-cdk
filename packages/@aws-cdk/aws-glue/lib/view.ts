import { Fn, Resource } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { IDatabase } from './database';
import { CfnTable } from './glue.generated';
import { Column, Schema } from './schema';
import { ITable } from './table';

/**
 * View props.
 *
 * @experimental
 */
export interface ViewProps {
  /**
   * Name of the table.
   */
  readonly tableName: string;

  /**
   * Description of the table.
   *
   * @default generated
   */
  readonly description?: string;

  /**
   * Database in which to store the table.
   */
  readonly database: IDatabase;

  /**
   * Columns of the table.
   */
  readonly columns: Column[];

  /**
   * Statement to run for this view.
   */
  readonly statement: string;

  /**
   * Placeholders to replace in the statement.
   *
   * @default - `database` resolves to the databaseName of {database} and can't be overridden.
   */
  readonly placeHolders?: { [key: string]: string };
}

// the mapping of athena create table to PrestoDB types
const prestoTypes: { [key: string]: string } = {
  [Schema.FLOAT.inputString]: 'real',
  [Schema.BIG_INT.inputString]: 'bigint',
  [Schema.INTEGER.inputString]: 'integer',
  [Schema.SMALL_INT.inputString]: 'smallint',
  [Schema.TINY_INT.inputString]: 'tinyint',
  [Schema.STRING.inputString]: 'varchar',
  [Schema.BINARY.inputString]: 'varbinary',
};

/**
 * A View in AWS Glue, implemented by an AWS Glue table.
 *
 * @experimental
 * @resource AWS::glue::Table
 */
export class View extends Resource implements ITable {
  /**
   * Name of this table.
   */
  public readonly tableName: string;

  /**
   * ARN of this table.
   */
  public readonly tableArn: string;

  /**
   * @summary Creates a Table construct that represents a view in the AWS Glue data catalogue.
   *
   * @param scope The scope creating construct (usually `this`).
   * @param id The construct's id.
   * @param props Properties.
   */
  constructor(scope: Construct, id: string, props: ViewProps) {
    super(scope, id, {
      physicalName: props.tableName,
    });

    const columns = props.columns;

    const placeHolders = {};
    if (props.placeHolders) {
      Object.assign(placeHolders, props.placeHolders);
    }
    Object.assign(placeHolders, { database: props.database.databaseName });

    const viewOriginalText = {
      originalSql: props.statement,
      catalog: 'awsdatacatalog',
      columns: this.renderColumns(columns, true),
      schema: '${database}',
    };

    const tableResource = new CfnTable(this, 'Table', {
      catalogId: props.database.catalogId,
      databaseName: props.database.databaseName,
      tableInput: {
        name: this.physicalName,
        description: props.description || `${props.tableName} generated by CDK`,
        parameters: { presto_view: true },
        storageDescriptor: {
          columns: this.renderColumns(columns),
          serdeInfo: {},
        },
        partitionKeys: [],
        tableType: 'VIRTUAL_VIEW',
        viewOriginalText:
          '/* Presto View: ' + Fn.base64(Fn.sub(JSON.stringify(viewOriginalText), placeHolders)) + ' */',
      },
    });
    this.node.defaultChild = tableResource;

    this.tableName = this.getResourceNameAttribute(tableResource.ref);
    this.tableArn = this.stack.formatArn({
      service: 'glue',
      resource: 'table',
      resourceName: `${props.database.databaseName}/${this.tableName}`,
    });
  }

  /**
   * Writes a json object used by glue from the list of columns.
   *
   * @param columns the columns
   * @param usePrestoTypes if `true`, replaces all data types with their presto types
   */
  private renderColumns(columns?: Column[], usePrestoTypes = false) {
    if (columns === undefined) {
      return undefined;
    }
    return columns.map((column) => {
      return {
        name: column.name,
        type:
          usePrestoTypes && column.type.inputString in prestoTypes
            ? prestoTypes[column.type.inputString]
            : column.type.inputString,
        comment: column.comment,
      };
    });
  }
}
