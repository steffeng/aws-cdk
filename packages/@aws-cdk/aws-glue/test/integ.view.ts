#!/usr/bin/env node
import * as fs from 'fs';
import * as path from 'path';
import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';
import * as glue from '../lib';

/*
 * Stack verification steps:
 * * 1. Run a query on the created view, use the databucket below as output:
 * *  a. aws athena start-query-execution --query-string "SELECT * FROM mydatabase.combined_view" --result-configuration OutputLocation=s3://test-databucketd8691f4e-12wylrvb09hdt
 * *
 * * 2. Take the query execution id from the output and check for "State": "SUCCEEDED"
 * *  a. aws athena get-query-execution --query-execution-id a58df117-9837-4667-96a3-253687350c96
 */

const app = new cdk.App();

const stack = new cdk.Stack(app, 'aws-cdk-glue');

const database = new glue.Database(stack, 'database', {
  databaseName: 'mydatabase',
});

const dataBucket = new s3.Bucket(stack, 'dataBucket');

const partitionKeys = [
  { name: 'year', type: glue.Schema.STRING },
  { name: 'month', type: glue.Schema.STRING },
  { name: 'day', type: glue.Schema.STRING },
  { name: 'hour', type: glue.Schema.STRING },
];

const sourceTable = new glue.Table(stack, 'sourceTable', {
  columns: [
    { name: 'date', type: glue.Schema.DATE },
    { name: 'time', type: glue.Schema.STRING },
    { name: 'location', type: glue.Schema.STRING },
    { name: 'bytes', type: glue.Schema.BIG_INT },
    { name: 'request_ip', type: glue.Schema.STRING },
    { name: 'method', type: glue.Schema.STRING },
    { name: 'host', type: glue.Schema.STRING },
    { name: 'uri', type: glue.Schema.STRING },
    { name: 'status', type: glue.Schema.INTEGER },
    { name: 'referrer', type: glue.Schema.STRING },
    { name: 'user_agent', type: glue.Schema.STRING },
    { name: 'query_string', type: glue.Schema.STRING },
    { name: 'cookie', type: glue.Schema.STRING },
    { name: 'result_type', type: glue.Schema.STRING },
    { name: 'request_id', type: glue.Schema.STRING },
    { name: 'host_header', type: glue.Schema.STRING },
    { name: 'request_protocol', type: glue.Schema.STRING },
    { name: 'request_bytes', type: glue.Schema.BIG_INT },
    { name: 'time_taken', type: glue.Schema.FLOAT },
    { name: 'xforwarded_for', type: glue.Schema.STRING },
    { name: 'ssl_protocol', type: glue.Schema.STRING },
    { name: 'ssl_cipher', type: glue.Schema.STRING },
    { name: 'response_result_type', type: glue.Schema.STRING },
    { name: 'http_version', type: glue.Schema.STRING },
    { name: 'fle_status', type: glue.Schema.STRING },
    { name: 'fle_encrypted_fields', type: glue.Schema.INTEGER },
    { name: 'c_port', type: glue.Schema.INTEGER },
    { name: 'time_to_first_byte', type: glue.Schema.FLOAT },
    { name: 'x_edge_detailed_result_type', type: glue.Schema.STRING },
    { name: 'sc_content_type', type: glue.Schema.STRING },
    { name: 'sc_content_len', type: glue.Schema.BIG_INT },
    { name: 'sc_range_start', type: glue.Schema.BIG_INT },
    { name: 'sc_range_end', type: glue.Schema.BIG_INT },
  ],
  partitionKeys: partitionKeys,
  database: database,
  tableName: 'partitioned_gz',
  description: 'Gzip logs delivered by Amazon CloudFront partitioned',
  s3Prefix: 'partitioned_gz',
  bucket: dataBucket,
  dataFormat: {
    outputFormat: glue.OutputFormat.HIVE_IGNORE_KEY_TEXT,
    inputFormat: glue.InputFormat.TEXT,
    serializationLibrary: glue.SerializationLibrary.LAZY_SIMPLE,
  },
});

const sourceCfnTable = sourceTable.node.defaultChild as glue.CfnTable;
sourceCfnTable.addOverride(
  'Properties.TableInput.Parameters.skip\\.header\\.line\\.count',
  '2',
);
sourceCfnTable.addOverride(
  'Properties.TableInput.StorageDescriptor.SerdeInfo.Parameters.field\\.delim',
  '\\t',
);
sourceCfnTable.addOverride(
  'Properties.TableInput.StorageDescriptor.SerdeInfo.Parameters.serialization\\.format',
  '\\t',
);

const targetTable = new glue.Table(stack, 'targetTable', {
  columns: sourceTable.columns,
  partitionKeys: sourceTable.partitionKeys,
  database: database,
  tableName: 'partitioned_parquet',
  description: 'Gzip logs delivered by Amazon CloudFront partitioned',
  s3Prefix: 'partitioned_parquet',
  bucket: dataBucket,
  dataFormat: {
    outputFormat: glue.OutputFormat.PARQUET,
    inputFormat: glue.InputFormat.PARQUET,
    serializationLibrary: glue.SerializationLibrary.PARQUET,
  },
});

const targetCfnTable = targetTable.node.defaultChild as glue.CfnTable;
targetCfnTable.addOverride(
  'Properties.TableInput.Parameters.parquet\\.compression',
  'snappy',
);

new glue.View(stack, 'combinedView', {
  columns: sourceTable.columns.concat(partitionKeys, {
    name: 'file',
    type: glue.Schema.STRING,
  }),
  database: database,
  tableName: 'combined_view',
  description: 'Gzip logs delivered by Amazon CloudFront partitioned',
  statement: fs
    .readFileSync(path.join(__dirname, 'combinedView.sql'))
    .toString(),
  placeHolders: {
    sourceTable: sourceTable.tableName,
    targetTable: targetTable.tableName,
  },
});

app.synth();
