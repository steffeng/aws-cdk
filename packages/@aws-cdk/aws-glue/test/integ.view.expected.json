{
  "Resources": {
    "databaseEBDE4557": {
      "Type": "AWS::Glue::Database",
      "Properties": {
        "CatalogId": {
          "Ref": "AWS::AccountId"
        },
        "DatabaseInput": {
          "Name": "mydatabase"
        }
      }
    },
    "dataBucketD8691F4E": {
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "sourceTable961262B0": {
      "Type": "AWS::Glue::Table",
      "Properties": {
        "CatalogId": {
          "Ref": "AWS::AccountId"
        },
        "DatabaseName": {
          "Ref": "databaseEBDE4557"
        },
        "TableInput": {
          "Description": "Gzip logs delivered by Amazon CloudFront partitioned",
          "Name": "partitioned_gz",
          "Parameters": {
            "has_encrypted_data": false,
            "skip.header.line.count": "2"
          },
          "PartitionKeys": [
            {
              "Name": "year",
              "Type": "string"
            },
            {
              "Name": "month",
              "Type": "string"
            },
            {
              "Name": "day",
              "Type": "string"
            },
            {
              "Name": "hour",
              "Type": "string"
            }
          ],
          "StorageDescriptor": {
            "Columns": [
              {
                "Name": "date",
                "Type": "date"
              },
              {
                "Name": "time",
                "Type": "string"
              },
              {
                "Name": "location",
                "Type": "string"
              },
              {
                "Name": "bytes",
                "Type": "bigint"
              },
              {
                "Name": "request_ip",
                "Type": "string"
              },
              {
                "Name": "method",
                "Type": "string"
              },
              {
                "Name": "host",
                "Type": "string"
              },
              {
                "Name": "uri",
                "Type": "string"
              },
              {
                "Name": "status",
                "Type": "int"
              },
              {
                "Name": "referrer",
                "Type": "string"
              },
              {
                "Name": "user_agent",
                "Type": "string"
              },
              {
                "Name": "query_string",
                "Type": "string"
              },
              {
                "Name": "cookie",
                "Type": "string"
              },
              {
                "Name": "result_type",
                "Type": "string"
              },
              {
                "Name": "request_id",
                "Type": "string"
              },
              {
                "Name": "host_header",
                "Type": "string"
              },
              {
                "Name": "request_protocol",
                "Type": "string"
              },
              {
                "Name": "request_bytes",
                "Type": "bigint"
              },
              {
                "Name": "time_taken",
                "Type": "float"
              },
              {
                "Name": "xforwarded_for",
                "Type": "string"
              },
              {
                "Name": "ssl_protocol",
                "Type": "string"
              },
              {
                "Name": "ssl_cipher",
                "Type": "string"
              },
              {
                "Name": "response_result_type",
                "Type": "string"
              },
              {
                "Name": "http_version",
                "Type": "string"
              },
              {
                "Name": "fle_status",
                "Type": "string"
              },
              {
                "Name": "fle_encrypted_fields",
                "Type": "int"
              },
              {
                "Name": "c_port",
                "Type": "int"
              },
              {
                "Name": "time_to_first_byte",
                "Type": "float"
              },
              {
                "Name": "x_edge_detailed_result_type",
                "Type": "string"
              },
              {
                "Name": "sc_content_type",
                "Type": "string"
              },
              {
                "Name": "sc_content_len",
                "Type": "bigint"
              },
              {
                "Name": "sc_range_start",
                "Type": "bigint"
              },
              {
                "Name": "sc_range_end",
                "Type": "bigint"
              }
            ],
            "Compressed": false,
            "InputFormat": "org.apache.hadoop.mapred.TextInputFormat",
            "Location": {
              "Fn::Join": [
                "",
                [
                  "s3://",
                  {
                    "Ref": "dataBucketD8691F4E"
                  },
                  "/partitioned_gz"
                ]
              ]
            },
            "OutputFormat": "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat",
            "SerdeInfo": {
              "Parameters": {
                "field.delim": "\\t",
                "serialization.format": "\\t"
              },
              "SerializationLibrary": "org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe"
            },
            "StoredAsSubDirectories": false
          },
          "TableType": "EXTERNAL_TABLE"
        }
      }
    },
    "targetTable65C350B8": {
      "Type": "AWS::Glue::Table",
      "Properties": {
        "CatalogId": {
          "Ref": "AWS::AccountId"
        },
        "DatabaseName": {
          "Ref": "databaseEBDE4557"
        },
        "TableInput": {
          "Description": "Gzip logs delivered by Amazon CloudFront partitioned",
          "Name": "partitioned_parquet",
          "Parameters": {
            "has_encrypted_data": false,
            "parquet.compression": "snappy"
          },
          "PartitionKeys": [
            {
              "Name": "year",
              "Type": "string"
            },
            {
              "Name": "month",
              "Type": "string"
            },
            {
              "Name": "day",
              "Type": "string"
            },
            {
              "Name": "hour",
              "Type": "string"
            }
          ],
          "StorageDescriptor": {
            "Columns": [
              {
                "Name": "date",
                "Type": "date"
              },
              {
                "Name": "time",
                "Type": "string"
              },
              {
                "Name": "location",
                "Type": "string"
              },
              {
                "Name": "bytes",
                "Type": "bigint"
              },
              {
                "Name": "request_ip",
                "Type": "string"
              },
              {
                "Name": "method",
                "Type": "string"
              },
              {
                "Name": "host",
                "Type": "string"
              },
              {
                "Name": "uri",
                "Type": "string"
              },
              {
                "Name": "status",
                "Type": "int"
              },
              {
                "Name": "referrer",
                "Type": "string"
              },
              {
                "Name": "user_agent",
                "Type": "string"
              },
              {
                "Name": "query_string",
                "Type": "string"
              },
              {
                "Name": "cookie",
                "Type": "string"
              },
              {
                "Name": "result_type",
                "Type": "string"
              },
              {
                "Name": "request_id",
                "Type": "string"
              },
              {
                "Name": "host_header",
                "Type": "string"
              },
              {
                "Name": "request_protocol",
                "Type": "string"
              },
              {
                "Name": "request_bytes",
                "Type": "bigint"
              },
              {
                "Name": "time_taken",
                "Type": "float"
              },
              {
                "Name": "xforwarded_for",
                "Type": "string"
              },
              {
                "Name": "ssl_protocol",
                "Type": "string"
              },
              {
                "Name": "ssl_cipher",
                "Type": "string"
              },
              {
                "Name": "response_result_type",
                "Type": "string"
              },
              {
                "Name": "http_version",
                "Type": "string"
              },
              {
                "Name": "fle_status",
                "Type": "string"
              },
              {
                "Name": "fle_encrypted_fields",
                "Type": "int"
              },
              {
                "Name": "c_port",
                "Type": "int"
              },
              {
                "Name": "time_to_first_byte",
                "Type": "float"
              },
              {
                "Name": "x_edge_detailed_result_type",
                "Type": "string"
              },
              {
                "Name": "sc_content_type",
                "Type": "string"
              },
              {
                "Name": "sc_content_len",
                "Type": "bigint"
              },
              {
                "Name": "sc_range_start",
                "Type": "bigint"
              },
              {
                "Name": "sc_range_end",
                "Type": "bigint"
              }
            ],
            "Compressed": false,
            "InputFormat": "org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat",
            "Location": {
              "Fn::Join": [
                "",
                [
                  "s3://",
                  {
                    "Ref": "dataBucketD8691F4E"
                  },
                  "/partitioned_parquet"
                ]
              ]
            },
            "OutputFormat": "org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat",
            "SerdeInfo": {
              "SerializationLibrary": "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe"
            },
            "StoredAsSubDirectories": false
          },
          "TableType": "EXTERNAL_TABLE"
        }
      }
    },
    "combinedViewTable71E67195": {
      "Type": "AWS::Glue::Table",
      "Properties": {
        "CatalogId": {
          "Ref": "AWS::AccountId"
        },
        "DatabaseName": {
          "Ref": "databaseEBDE4557"
        },
        "TableInput": {
          "Description": "Gzip logs delivered by Amazon CloudFront partitioned",
          "Name": "combined_view",
          "Parameters": {
            "presto_view": true
          },
          "PartitionKeys": [],
          "StorageDescriptor": {
            "Columns": [
              {
                "Name": "date",
                "Type": "date"
              },
              {
                "Name": "time",
                "Type": "string"
              },
              {
                "Name": "location",
                "Type": "string"
              },
              {
                "Name": "bytes",
                "Type": "bigint"
              },
              {
                "Name": "request_ip",
                "Type": "string"
              },
              {
                "Name": "method",
                "Type": "string"
              },
              {
                "Name": "host",
                "Type": "string"
              },
              {
                "Name": "uri",
                "Type": "string"
              },
              {
                "Name": "status",
                "Type": "int"
              },
              {
                "Name": "referrer",
                "Type": "string"
              },
              {
                "Name": "user_agent",
                "Type": "string"
              },
              {
                "Name": "query_string",
                "Type": "string"
              },
              {
                "Name": "cookie",
                "Type": "string"
              },
              {
                "Name": "result_type",
                "Type": "string"
              },
              {
                "Name": "request_id",
                "Type": "string"
              },
              {
                "Name": "host_header",
                "Type": "string"
              },
              {
                "Name": "request_protocol",
                "Type": "string"
              },
              {
                "Name": "request_bytes",
                "Type": "bigint"
              },
              {
                "Name": "time_taken",
                "Type": "float"
              },
              {
                "Name": "xforwarded_for",
                "Type": "string"
              },
              {
                "Name": "ssl_protocol",
                "Type": "string"
              },
              {
                "Name": "ssl_cipher",
                "Type": "string"
              },
              {
                "Name": "response_result_type",
                "Type": "string"
              },
              {
                "Name": "http_version",
                "Type": "string"
              },
              {
                "Name": "fle_status",
                "Type": "string"
              },
              {
                "Name": "fle_encrypted_fields",
                "Type": "int"
              },
              {
                "Name": "c_port",
                "Type": "int"
              },
              {
                "Name": "time_to_first_byte",
                "Type": "float"
              },
              {
                "Name": "x_edge_detailed_result_type",
                "Type": "string"
              },
              {
                "Name": "sc_content_type",
                "Type": "string"
              },
              {
                "Name": "sc_content_len",
                "Type": "bigint"
              },
              {
                "Name": "sc_range_start",
                "Type": "bigint"
              },
              {
                "Name": "sc_range_end",
                "Type": "bigint"
              },
              {
                "Name": "year",
                "Type": "string"
              },
              {
                "Name": "month",
                "Type": "string"
              },
              {
                "Name": "day",
                "Type": "string"
              },
              {
                "Name": "hour",
                "Type": "string"
              },
              {
                "Name": "file",
                "Type": "string"
              }
            ],
            "SerdeInfo": {}
          },
          "TableType": "VIRTUAL_VIEW",
          "ViewOriginalText": {
            "Fn::Join": [
              "",
              [
                "/* Presto View: ",
                {
                  "Fn::Base64": {
                    "Fn::Sub": [
                      "{\"originalSql\":\"SELECT\\n  *\\n, \\\"$path\\\" \\\"file\\\" \\nFROM\\n  ${sourceTable}\\nWHERE (\\\"concat\\\"(\\\"year\\\", \\\"month\\\", \\\"day\\\", \\\"hour\\\") >= \\\"date_format\\\"(\\\"date_trunc\\\"('hour', ((current_timestamp - INTERVAL  '15' MINUTE) - INTERVAL  '1' HOUR)), '%Y%m%d%H'))\\nUNION ALL SELECT\\n  *\\n, \\\"$path\\\" \\\"file\\\"\\nFROM\\n  ${targetTable}\\nWHERE (\\\"concat\\\"(\\\"year\\\", \\\"month\\\", \\\"day\\\", \\\"hour\\\") < \\\"date_format\\\"(\\\"date_trunc\\\"('hour', ((current_timestamp - INTERVAL  '15' MINUTE) - INTERVAL  '1' HOUR)), '%Y%m%d%H'))\\n\",\"catalog\":\"awsdatacatalog\",\"columns\":[{\"name\":\"date\",\"type\":\"date\"},{\"name\":\"time\",\"type\":\"varchar\"},{\"name\":\"location\",\"type\":\"varchar\"},{\"name\":\"bytes\",\"type\":\"bigint\"},{\"name\":\"request_ip\",\"type\":\"varchar\"},{\"name\":\"method\",\"type\":\"varchar\"},{\"name\":\"host\",\"type\":\"varchar\"},{\"name\":\"uri\",\"type\":\"varchar\"},{\"name\":\"status\",\"type\":\"integer\"},{\"name\":\"referrer\",\"type\":\"varchar\"},{\"name\":\"user_agent\",\"type\":\"varchar\"},{\"name\":\"query_string\",\"type\":\"varchar\"},{\"name\":\"cookie\",\"type\":\"varchar\"},{\"name\":\"result_type\",\"type\":\"varchar\"},{\"name\":\"request_id\",\"type\":\"varchar\"},{\"name\":\"host_header\",\"type\":\"varchar\"},{\"name\":\"request_protocol\",\"type\":\"varchar\"},{\"name\":\"request_bytes\",\"type\":\"bigint\"},{\"name\":\"time_taken\",\"type\":\"real\"},{\"name\":\"xforwarded_for\",\"type\":\"varchar\"},{\"name\":\"ssl_protocol\",\"type\":\"varchar\"},{\"name\":\"ssl_cipher\",\"type\":\"varchar\"},{\"name\":\"response_result_type\",\"type\":\"varchar\"},{\"name\":\"http_version\",\"type\":\"varchar\"},{\"name\":\"fle_status\",\"type\":\"varchar\"},{\"name\":\"fle_encrypted_fields\",\"type\":\"integer\"},{\"name\":\"c_port\",\"type\":\"integer\"},{\"name\":\"time_to_first_byte\",\"type\":\"real\"},{\"name\":\"x_edge_detailed_result_type\",\"type\":\"varchar\"},{\"name\":\"sc_content_type\",\"type\":\"varchar\"},{\"name\":\"sc_content_len\",\"type\":\"bigint\"},{\"name\":\"sc_range_start\",\"type\":\"bigint\"},{\"name\":\"sc_range_end\",\"type\":\"bigint\"},{\"name\":\"year\",\"type\":\"varchar\"},{\"name\":\"month\",\"type\":\"varchar\"},{\"name\":\"day\",\"type\":\"varchar\"},{\"name\":\"hour\",\"type\":\"varchar\"},{\"name\":\"file\",\"type\":\"varchar\"}],\"schema\":\"${database}\"}",
                      {
                        "sourceTable": {
                          "Ref": "sourceTable961262B0"
                        },
                        "targetTable": {
                          "Ref": "targetTable65C350B8"
                        },
                        "database": {
                          "Ref": "databaseEBDE4557"
                        }
                      }
                    ]
                  }
                },
                " */"
              ]
            ]
          }
        }
      }
    }
  }
}